### ISSUE TITLE
fromJson() not called and default constructor assigns incorrect defaults

### ISSUE THREAD

**Issue by rtedin** [immutables/immutables#226]
Given the following example interface:

```java
@Value.Immutable
@Value.Style (
        implementationNestedInBuilder = true,
        visibility = Value.Style.ImplementationVisibility.PACKAGE,
        builderVisibility = Value.Style.BuilderVisibility.PUBLIC,
        privateNoargConstructor = true
)
@JsonDeserialize(as = ExampleBuilder.ImmutableExample.class)
public interface Example {

    List<String> getStrings();
}
```

The following test does not pass:

```java
public class ExampleTest {

    @Test
    public void testGetStringsFromJson() throws Exception {

        ObjectMapper objectMapper = new ObjectMapper();

        String json = "{}"; // Does not pass he test.
        // String json = "{\"strings\":null}"; // Does not pass he test.
        // String json = "{\"strings\":[\"asd\"]}"; // Passes the test.

        Example example = objectMapper.readValue(json, Example.class);

        MatcherAssert.assertThat(example.getStrings(), CoreMatchers.notNullValue());
    }
}
```

The problem seems to be that Jackson uses the default constructor of `ExampleBuilder.ImmutableExample` instead of the `fromJson` method. But the default constructor assigns `null` to the list, breaking what we thought was the contract of `Example`. I.e. if we do not annotate `getStrings()` with for instance with `@Nullable`, then the immutable implementation would take care of assigning meaningful defaults. In this case, if the default constructor would be:

```java
private ImmutableExample() {
      this.strings = createUnmodifiableList(false, new ArrayList<String>());
}
```

then the test would pass.

Note that omitting the `strings` field in the JSON (as part of a bigger object, this is just a minimal example) is a valid use case for us, for instance for saving bandwidth when transferring big objects. We have this issue also with maps. We are using Jackson 2.4.3.

Many thanks for the great work. We're using Immutables more and more thanks to you!

**Comment by elucash** [comment]
Thank you for the report! I'm curious why do you need private constructor functionality. It was created for "exotic" reflective frameworks, and consistency is not enforced there, so I would not try to use `privateNoargConstructor` in any way.  For some reason `@JsonCreator` does not work as expected here. I tried to create some contrived cases, but interestingly `@JsonCreator` work in similar setup. Maybe even we could report bug to once we can narrow down problem. As a workaround you might try to different structural configuration/styles. I'll get back to this in a day or two to try to find specific bug/workaround in Immutables or Jackson.

**Comment by rtedin** [comment]
If we don't use `privateNoargConstructor` we get the following error when executing the test:

```
com.fasterxml.jackson.databind.JsonMappingException: No suitable constructor found for type [simple type, class test.ExampleBuilder$ImmutableExample]: can not instantiate from JSON object (need to add/enable type information?)
 at [Source: {}; line: 1, column: 2]
```

**Comment by elucash** [comment]
Yes, but `@JsonCreator` should be picked up – thats the bug I was referring to.
If you switch class to public, it will most probably work that way.

```java
@Value.Style (
        implementationNestedInBuilder = true,
        visibility = Value.Style.ImplementationVisibility.PUBLIC,
        builderVisibility = Value.Style.BuilderVisibility.PUBLIC,
```
In any way, this need to be solved somehow, but I would not look at `privateNoargConstructor` as a "solution" for this

**Comment by rtedin** [comment]
You're right, it works with public visibility. I agree that this should be solved to allow other visibility values, if Jackson supports it.

**Comment by rtedin** [comment]
Any news about this?

**Comment by elucash** [comment]
sorry, no news about it yet. I'll take another look at it shortly, and I really hoping that this is not blocking you

**Comment by rtedin** [comment]
No, it's not blocking us since we have a work around. Would be still nice to have this solved so that we could simplify our code though. Many thanks for your efforts!

**Comment by elucash** [comment]
Good news, we found the cause and fixed generation of json-creator method a bit. Will be released in 2.1.9, next week, probably

**Comment by elucash** [comment]
released in 2.1.9

**Comment by rtedin** [comment]
Thanks! We will try this in a short time.
